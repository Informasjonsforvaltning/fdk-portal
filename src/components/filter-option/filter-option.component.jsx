import React from 'react';
import PropTypes from 'prop-types';
import _capitalize from 'lodash/capitalize';

import localization from '../../lib/localization';
import { getTranslateText } from '../../lib/translateText';

export const FilterOption = props => {
  const {
    itemKey,
    value,
    labelRaw,
    label,
    count,
    onClick,
    active,
    themesItems,
    displayClass
  } = props;
  const optionLabel = labelRaw || label;

  let textLabel;
  // if themes, then choose text from themes array, else choose label from localization-file.
  if (themesItems) {
    if (themesItems[`${label}`]) {
      textLabel = getTranslateText(themesItems[`${label}`].title);
    } else {
      textLabel = localization.unknown;
    }
  } else if (
    optionLabel.toUpperCase() === 'UKJENT' ||
    optionLabel.toUpperCase() === 'MISSING'
  ) {
    textLabel = localization.unknown;
  } else {
    textLabel = localization[optionLabel] || optionLabel;
  }
  if (textLabel && textLabel === textLabel.toUpperCase()) {
    textLabel =
      localization[textLabel.toLowerCase()] || `${_capitalize(textLabel)}`;
  }

  const id = encodeURIComponent(itemKey + value);

  let inputRef;

  return (
    <div className={`checkbox ${displayClass}`}>
      <label
        className="checkbox_label"
        onKeyPress={() => {
          inputRef.click();
        }}
        tabIndex="0"
        role="button"
        htmlFor={id}
      >
        <input
          ref={input => {
            inputRef = input;
          }}
          type="checkbox"
          id={id}
          tabIndex="-1"
          checked={active}
          onChange={e => onClick(e)}
          className="list-group-item fdk-label fdk-label-default"
          value={value}
        />
        <span className="checkbox-replacement" />
        {textLabel} ({count})
      </label>
    </div>
  );
};

FilterOption.defaultProps = {
  value: null,
  labelRaw: null,
  label: null,
  count: null,
  active: null,
  themesItems: null,
  displayClass: null
};

FilterOption.propTypes = {
  itemKey: PropTypes.number.isRequired,
  value: PropTypes.string,
  labelRaw: PropTypes.string,
  label: PropTypes.string,
  count: PropTypes.number,
  onClick: PropTypes.func.isRequired,
  active: PropTypes.bool,
  themesItems: PropTypes.object,
  displayClass: PropTypes.string
};
