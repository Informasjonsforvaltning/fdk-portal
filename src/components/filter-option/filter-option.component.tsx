import React from 'react';
import type { FC, ChangeEventHandler } from 'react';
import _capitalize from 'lodash/capitalize';

import localization from '../../lib/localization';
import { getTranslateText } from '../../lib/translateText';

interface Props {
  itemKey: number;
  value?: string;
  labelRaw?: string;
  label?: string;
  count?: number;
  onClick: ChangeEventHandler;
  active?: boolean;
  referenceDataItems?: Record<string, any>;
  displayClass?: string;
  capitalize?: boolean;
}

export const FilterOption: FC<Props> = ({
  itemKey,
  value,
  labelRaw,
  label,
  count,
  onClick,
  active,
  referenceDataItems,
  displayClass,
  capitalize = true
}) => {
  const optionLabel = labelRaw || label;

  let textLabel;
  // if themes, then choose text from themes array, else choose label from localization-file.
  if (referenceDataItems) {
    if (referenceDataItems[`${label}`]) {
      textLabel =
        getTranslateText(referenceDataItems[`${label}`].label) ||
        getTranslateText(referenceDataItems[`${label}`].name);
    } else {
      textLabel = localization.unknown;
    }
  } else if (
    optionLabel?.toUpperCase() === 'UKJENT' ||
    optionLabel?.toUpperCase() === 'MISSING'
  ) {
    textLabel = localization.unknown;
  } else if (optionLabel?.toUpperCase()?.startsWith('/ANNET')) {
    textLabel = _capitalize(
      optionLabel.substr(optionLabel.lastIndexOf('/') + 1, optionLabel.length)
    );
  } else {
    textLabel = (optionLabel && localization[optionLabel]) || optionLabel;
  }
  if (textLabel && textLabel === textLabel.toUpperCase()) {
    textLabel =
      localization[textLabel.toLowerCase()] ||
      (capitalize ? `${_capitalize(textLabel)}` : textLabel);
  }

  const id = encodeURIComponent(itemKey + (value ?? ''));

  return (
    <div className={`checkbox ${displayClass}`}>
      <label
        className='checkbox_label'
        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role
        htmlFor={id}
      >
        <input
          type='checkbox'
          id={id}
          checked={active}
          onChange={onClick}
          className='list-group-item fdk-label fdk-label-default'
          value={value}
        />
        <span className='checkbox-replacement' />
        {textLabel} {count ? `(${count})` : ''}
      </label>
    </div>
  );
};
