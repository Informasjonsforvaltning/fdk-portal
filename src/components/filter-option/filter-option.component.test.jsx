import React from 'react';
import { shallow } from 'enzyme';
import { FilterOption } from './filter-option.component';
import filterOptions from '../../../test/fixtures/filterOptions';

test('should render FilterOption correctly', () => {
  const onClick = jest.fn();
  const wrapper = shallow(
    <FilterOption
      itemKey={filterOptions[0].itemKey}
      value={filterOptions[0].value}
      label={filterOptions[0].label}
      count={filterOptions[0].count}
      onClick={onClick}
      active={filterOptions[0].active}
      themesItems={filterOptions[0].themesItems}
      displayClass={filterOptions[0].displayClass}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render FilterOption correctly with no corresponding theme', () => {
  const onClick = jest.fn();
  const wrapper = shallow(
    <FilterOption
      itemKey={filterOptions[1].itemKey}
      value={filterOptions[1].value}
      label={filterOptions[1].label}
      count={filterOptions[1].count}
      onClick={onClick}
      active={filterOptions[1].active}
      themesItems={filterOptions[1].themesItems}
      displayClass={filterOptions[1].displayClass}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render FilterOption correctly with no theme', () => {
  const onClick = jest.fn();
  const wrapper = shallow(
    <FilterOption
      itemKey={filterOptions[2].itemKey}
      value={filterOptions[2].value}
      label={filterOptions[2].label}
      count={filterOptions[2].count}
      onClick={onClick}
      active={filterOptions[2].active}
      themesItems={filterOptions[2].themesItems}
      displayClass={filterOptions[2].displayClass}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should handle onChange', () => {
  const onClick = jest.fn();
  const wrapper = shallow(
    <FilterOption
      itemKey={filterOptions[2].itemKey}
      value={filterOptions[2].value}
      label={filterOptions[2].label}
      count={filterOptions[2].count}
      onClick={onClick}
      active={filterOptions[2].active}
      themesItems={filterOptions[2].themesItems}
      displayClass={filterOptions[2].displayClass}
    />
  );
  wrapper.find('input.list-group-item').prop('onChange')(filterOptions[1]);
  expect(onClick).toHaveBeenLastCalledWith(filterOptions[1]);
});
