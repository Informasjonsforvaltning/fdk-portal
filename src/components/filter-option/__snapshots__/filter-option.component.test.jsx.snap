// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FilterOption correctly 1`] = `
<div
  className="checkbox null"
>
  <label
    className="checkbox_label"
    htmlFor="4AGRI"
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <input
      checked={false}
      className="list-group-item fdk-label fdk-label-default"
      id="4AGRI"
      onChange={[Function]}
      tabIndex="-1"
      type="checkbox"
      value="AGRI"
    />
    <span
      className="checkbox-replacement"
    />
    Agriculture, fisheries, forestry and food
     (
    18
    )
  </label>
</div>
`;

exports[`should render FilterOption correctly with no corresponding theme 1`] = `
<div
  className="checkbox null"
>
  <label
    className="checkbox_label"
    htmlFor="4JUST"
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <input
      checked={false}
      className="list-group-item fdk-label fdk-label-default"
      id="4JUST"
      onChange={[Function]}
      tabIndex="-1"
      type="checkbox"
      value="JUST"
    />
    <span
      className="checkbox-replacement"
    />
    Unknown
     (
    18
    )
  </label>
</div>
`;

exports[`should render FilterOption correctly with no theme 1`] = `
<div
  className="checkbox null"
>
  <label
    className="checkbox_label"
    htmlFor="0PUBLIC"
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <input
      checked={false}
      className="list-group-item fdk-label fdk-label-default"
      id="0PUBLIC"
      onChange={[Function]}
      tabIndex="-1"
      type="checkbox"
      value="PUBLIC"
    />
    <span
      className="checkbox-replacement"
    />
    Public
     (
    463
    )
  </label>
</div>
`;
